### PR TITLE
 #169817550 Manager can get notification on new trip

### PR DESCRIPTION
**Pull request by victor-abz**
Manager can opt-in to certain notification mode(email / in-app)
Manager can opt-out to certain notification mode

[Starts #169817550]
#### What does this PR do?
User can get notification
#### Description of Task to be completed?
- Manager can opt-in to certain notification mode(email / in-app)
- Manager can opt-out to certain notification mode
- Manager can get notification by preferred mode when a new trip created
#### How should this be manually tested?
go to terminal and do `git clone https://github.com/andela/the_spinners-backend`
switch to a branch by `git checkout ft-notification-on-new-travel-169817550`
run `npm run dev`
###### To set preferred mode
PUT`/api/user/notification/set/email-notification` to set receive notification by email
PUT `/api/user/notification/set/in-app-notification` to receive notification in-app
###### To remove preferred mode
PUT `/api/users/notification/unset/email-notification` to unsubscribe from receiving notification by email
PUT `/api/user/notification/unset/in-app-notification` to unsubscribe from notification in-app
###### To receive sample notification
create a new trip by visiting POST `/api/one-way-trips`, you should get a notification in email.
to get all notification for a logged in user visit `/api/user/notification`

more on testing can be seen in our documentation [https://snippers-backend-stage.herokuapp.com/docs](https://snippers-backend-stage.herokuapp.com/docs)
#### What are the relevant pivotal tracker stories?
#169817550 
#### Screenshots (if appropriate)
Screen Shot 2020-01-01 at 23 45 30

#### Questions:

### CODE REVIEW

File: src/helpers/notifications/trip-notification.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        manager
+      } = await this.findManagerToNotify(request);
+      const notificationMessage = `New ${request.tripType} trip have been Requested by ${firstName} ${lastName}`;
+      const inAppNotification = await NotificationService.createNotification({
+        requestId: request.id,
+        notificationMessage,
+        type: 'new_request',
+        userId: id,
+      });
+      delete inAppNotification.dataValues.isRead;
```
Comment (higustave-ops) [inline src/helpers/notifications/trip-notification.js]: Can you use `rest param destructuring` instead of `delete`

File: src/services/send-email.service.js [hunk 2]
```diff
@@ -7,12 +7,11 @@ import JwtService from './jwt.service';
 class SendEmailService {
   /**
    * @param {email} email user email
-   * @param {token} token user token
    * @param {subject} emailSubject subject of the email
    * @param {body} emailBody description of the email
    * @returns {sendMessage} this function send an email
    */
-  static sendGridEmail(email, token, emailSubject, emailBody) {
```
Comment (higustave-ops) [inline src/services/send-email.service.js:15]: Is our features going to work as they used to, because this method was for all of us and now, you removed the token `arg`. Need explanation. Thanks

Comment (victor-abz) [reply src/services/send-email.service.js:15]: the token is not used anywhere in the body. s

Comment (higustave-ops) [reply src/services/send-email.service.js:15]: Did you check that other features that were depending on this method are working. For example mine of user account email verification? The email sent should contain a token> Try with your email and see if it works then will be ok. Thanks

**Review by higustave-ops — APPROVED** [review]
LGTM

File: src/helpers/validateErrorHandler.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import ResponseService from '../services/response.service';
+
+const validateErrorHandler = (schema, body, res, next) => {
```
Comment (erickyvand) [inline src/helpers/validateErrorHandler.js]: Why did you put validation in the helpers folder while we have a validations folder?

Comment (victor-abz) [reply src/helpers/validateErrorHandler.js]: This is a helper function which brings all errors from joi at once. I saw we all had to copy it and that would cause duplicates as code grows. I call it to generate the error without rewrite it.

**Comment by erickyvand** [comment]
LGTM

**Review by erickyvand — APPROVED** [review]

File: src/app.js [hunk 4]
```diff
@@ -3,14 +3,19 @@ import express from 'express';
 import morgan from 'morgan';
 import passport from 'passport';
 import routes from './routes/index';
+import NotificationListeners from './helpers/notifications/index';
 
 const app = express();
 app.use(morgan('dev'));
+app.use(express.static(`${__dirname}/public/`));
```
Comment (adesege) [inline src/app.js]: Is there a reason for adding this line?

File: src/controllers/notification.controller.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    if (req.option === 'inAppNotification') { // If request is inAppNotification
+      await PreferenceService
+        .updatePreference({ userId: req.userData.id }, { inAppNotification: req.value });
+      ResponseService.setSuccess(200, `Successfully updated your ${req.option} preferences.`);
+      return ResponseService.send(res);
+    }
+    // If request is emailNotification
+    await PreferenceService
+      .updatePreference({ userId: req.userData.id }, { emailNotification: req.value });
+    ResponseService.setSuccess(200, `Successfully updated your ${req.option} preferences.`);
```
Comment (adesege) [inline src/controllers/notification.controller.js]: `req.option` does not have a user-friendly name. Use a user-friendly name instead

File: src/helpers/notifications/trip-notification.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import emitter from '../eventEmmiters/emitter';
+import NotificationService from '../../services/notification.service';
+import UserService from '../../services/user.service';
+import app from '../eventEmmiters/socket';
+import tripEmailBody from '../mails/trip-notification.email';
+import SendEmailService from '../../services/send-email.service';
+import PreferenceService from '../../services/preference.service';
+
+/** Class representing a Notifications . */
+class TripNotification {
```
Comment (adesege) [inline src/helpers/notifications/trip-notification.js]: This should move to the services directory instead.

File: src/helpers/validateErrorHandler.js (hunk 3 above)
Comment (adesege) [inline src/helpers/validateErrorHandler.js]: `transformErrorHandler` is more succinct than `validateErrorHandler`.

File: src/helpers/validateErrorHandler.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  const { error } = schema.validate(body);
+
+  if (error) {
+    const errors = [];
+    const { details } = error;
+    details.forEach(({ message }) => {
```
Comment (adesege) [inline src/helpers/validateErrorHandler.js]: Use .map instead.

File: src/middlewares/auth.middleware.js [hunk 8]
```diff
@@ -11,6 +11,14 @@ const authMiddleware = {
     }
     next();
   },
+  doesHaveManager: async (req, res, next) => {
```
Comment (adesege) [inline src/middlewares/auth.middleware.js]: `doesHaveManager` is not a good function name. Ensure your function name has a verb and clearly describes what the function does.

File: src/middlewares/notification.middleware.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import PreferenceService from '../services/preference.service';
+import ResponseService from '../services/response.service';
+
+const doesOptionExist = async (req, res, next) => {
```
Comment (adesege) [inline src/middlewares/notification.middleware.js]: `doesOptionExist` is not a good function name. Ensure your function name has a verb and clearly describes what the function does.

File: src/middlewares/notification.middleware.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  const prefs = await PreferenceService.findPreferences({ userId: req.userData.id });
+  const { dataValues } = prefs;
+  Object.keys(dataValues)
+    .forEach((index) => {
```
Comment (adesege) [inline src/middlewares/notification.middleware.js]: You can use a combination of filter and map functions here.

File: src/migrations/20200103080032-create-preferences.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.INTEGER
+    },
+    userId: {
+      type: Sequelize.INTEGER
+    },
+    emailNotification: {
```
Comment (adesege) [inline src/migrations/20200103080032-create-preferences.js]: Variables that contain a boolean as a value should be prefixed with either `is` or `has`

File: src/routes/index.js [hunk 12]
```diff
@@ -2,11 +2,13 @@ import express from 'express';
...
 import tripRoute from './trip.route';
 import profileSettingsRoute from './profile-settings.route';
+import notificationRoute from './notification.route';
 
 const app = express();
 
 app.use('/api/auth', authRoute);
 app.use('/api/', tripRoute);
 app.use('/api/', profileSettingsRoute);
+app.use('/api/', notificationRoute);
```
Comment (adesege) [inline src/routes/index.js]: The context can be scoped to `/api/notifications`

**Review by adesege — CHANGES_REQUESTED** [review]

File: src/helpers/notifications/trip-notification.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      const data = await NotificationService.createNotification({
+        requestId: request.id,
+        notificationMessage,
+        type: 'new_request',
+        userId: id,
+      });
+      const { isRead, updateAt, createdAt, ...notificationData } = data.dataValues;
+
+      // email nofitication
+      if (emailNotification === true) {
```
Comment (adesege) [inline src/helpers/notifications/trip-notification.js]: Add a test case for this line and line 67

File: src/routes/index.js (hunk 12 above)
Comment (victor-abz) [reply src/routes/index.js]: Yeah, it is /api/notifications except that the last part `/notification` is in the notification route file.

File: .vscode/settings.json [hunk 14]
```diff
@@ -0,0 +1,3 @@
+{
+    "liveServer.settings.port": 5501
```
Comment (adesege) [inline .vscode/settings.json]: You don't need to push this file.

File: index.html [hunk 15]
```diff
@@ -0,0 +1,62 @@
+<!DOCTYPE html>
```
Comment (adesege) [inline index.html]: What does this HTML file do?

File: src/controllers/notification.controller.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+  /**
+   * @param  {object} req - Request object
+   *  @param {object} res - Response object
+   *  @returns {object} response
+   * */
+  static async setNotificationMode(req, res) {
+    if (req.preferenceMode === 'isInAppNotification') { // If request is isInAppNotification
+      await PreferenceService
+        .updatePreference({ userId: req.userData.id }, { isInAppNotification: req.value });
+      ResponseService.setSuccess(200, `Successfully updated your ${req.preferenceMode} preferences.`);
```
Comment (adesege) [inline src/controllers/notification.controller.js]: Regarding my previous feedback, `req.preferenceMode` is going to output one of `isInAppNotification` or `isEmailNotification` and that is not user-friendly.

A non-technical user does not know what `isInAppNotification` stands for, for example.

File: src/helpers/notifications/script.js [hunk 17]
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-console */
```
Comment (adesege) [inline src/helpers/notifications/script.js]: Do not disable any eslint rule

File: src/helpers/notifications/script.js [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+/* eslint-disable no-undef */
+/* eslint-disable no-alert */
+const socket = io('http://localhost:3000');
```
Comment (adesege) [inline src/helpers/notifications/script.js]: Do not hardcode the socket host. This will not work in production.

File: src/helpers/transformErrorHandler.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
+import ResponseService from '../services/response.service';
+
+const transformErrorHandler = (schema, body, res, next) => {
+  const { error } = schema.validate(body);
+
+  if (error) {
+    const errors = [];
+    const { details } = error;
+    details.map(({ message }) => errors.push(message.split('"').join(' ')));
```
Comment (adesege) [inline src/helpers/transformErrorHandler.js]: .map returns a new array. Assign a variable to the function instead of pushing to a new array.

File: src/middlewares/notification.middleware.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
+import PreferenceService from '../services/preference.service';
+import ResponseService from '../services/response.service';
+
+const checkIfOptionExists = async (req, res, next) => {
+  const prefs = await PreferenceService.findPreferences({ userId: req.userData.id });
+  const { dataValues } = prefs;
+
+  const isValueSet = Object.keys(dataValues)
+    .filter((key) => key === req.preferenceMode)
```
Comment (adesege) [inline src/middlewares/notification.middleware.js]: A user can only have one preference so there is no need to filter and map. Use sequelize find method to get only one preference from the database then do your check.

File: src/models/notification.js [hunk 21]
```diff
@@ -0,0 +1,19 @@
+export default (sequelize, DataTypes) => {
+  const Notifications = sequelize.define('Notifications', {
+    userId: DataTypes.INTEGER,
+    notificationMessage: {
```
Comment (adesege) [inline src/models/notification.js]: The notification prefix is not needed.

File: src/models/trip.js [hunk 22]
```diff
@@ -15,5 +18,9 @@ export default (sequelize, DataTypes) => {
     Trip.belongsTo(models.Location, { foreignKey: 'originId', targetKey: 'id' });
     Trip.belongsTo(models.Location, { foreignKey: 'destinationId', targetKey: 'id' });
   };
+  Trip.afterCreate(({ dataValues }) => {
+    emitter.emit('request-created', dataValues);
+    // TripNotification.sendTripNotification(dataValues);
```
Comment (adesege) [inline src/models/trip.js]: Do not leave commented code in your application. Ensure you remove all commented codes

File: src/seeders/20200106071439-demo-users.js [hunk 23]
```diff
@@ -0,0 +1,44 @@
+import BcryptService from '../services/bcrypt.service';
+
+const passwordHash = BcryptService.hashPassword('Pass1234');
+
+const demoUsers = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert('Users', [{
+    id: 31,
+    firstName: 'Manager',
```
Comment (adesege) [inline src/seeders/20200106071439-demo-users.js]: Use Faker to generate random data

File: src/seeders/20200106072811-users-notification-mode.js [hunk 24]
```diff
@@ -0,0 +1,26 @@
+const demoPreferenceMode = {
```
Comment (adesege) [inline src/seeders/20200106072811-users-notification-mode.js]: Is there really a need to generate seeders for this task?

File: src/tests/fixtures/users.fixture.js [hunk 25]
```diff
@@ -78,6 +84,22 @@ export const createUser = {
   createdAt: new Date(),
   updatedAt: new Date(),
 };
+export const userWithoutManager = {
+  id: 29,
```
Comment (adesege) [inline src/tests/fixtures/users.fixture.js]: ids should be randomly generated

File: src/tests/notification/notification.test.js [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+      .send({ isInAppNotification: false })
+      .end((err, res) => {
+        expect(res).to.have.status(200);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('message')
+          .that.contain('Successfully updated your isInAppNotification preferences.');
+        done(err);
+      });
+  });
+  it('Should not allow wrong notification mode', (done) => {
```
Comment (adesege) [inline src/tests/notification/notification.test.js]: This test case is duplicated

File: src/tests/notification/notification.test.js [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+      .send({ ...newTrip, travelDate: '2020-05-23' })
+      .end((err, res) => {
+        expect(res).to.have.status(401);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('message')
+          .that.contain('You don\'t have manager assigned yet. Please contact Admin');
+        done(err);
+      });
+  });
+  it('Should not accept requests with no manager', (done) => {
```
Comment (adesege) [inline src/tests/notification/notification.test.js:83]: This test case is duplicated

File: src/tests/trip/multi-city-trip.test.js [hunk 28]
```diff
@@ -120,7 +120,7 @@ describe('Test multi city trip:', () => {
     chai.request(app)
       .post('/api/multi-city-trips')
       .set('Authorization', loggedInToken)
-      .send([...multiCitytrip, { ...multiCitytrip[0], destinationId: 3 }])
+      .send([...multiCitytrip, { ...multiCitytrip[0], destinationId: 100 }])
```
Comment (adesege) [inline src/tests/trip/multi-city-trip.test.js]: `destinationId` should be randomly generated.

File: src/validations/notification.validation.js [hunk 29]
```diff
@@ -0,0 +1,21 @@
+import ResponseService from '../services/response.service';
+
+const validateNotification = (req, res, next) => {
+  if (Object.keys(req.body).length !== 1) {
```
Comment (adesege) [inline src/validations/notification.validation.js]: Use Joi to handle the validation instead

**Review by adesege — CHANGES_REQUESTED** [review]
Some feedback do not only apply to the line it was given. Ensure you go through your code before requesting for another review.

File: src/services/notification.service.js [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  /**
+   * @param {object} receiver
+   * @param {object} notification
+   * @return {function} send notification to connected client
+   */
+  static async sendNotifications(receiver, notification) {
+    // Ensure notification is an object which contain notificationMessage
+    if (!users[receiver]) return;
```
Comment (adesege) [inline src/services/notification.service.js]: Write test for this line

File: src/services/trip-notification.service.js [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+      const data = await NotificationService.createNotification({
+        requestId: request.id,
+        notificationMessage,
+        type: 'new_request',
+        userId: id,
+      });
+      const { isRead, updateAt, createdAt, ...notificationData } = data.dataValues;
+
+      // email nofitication
+      if (managerPreferences.isEmailNotification === true) {
```
Comment (adesege) [inline src/services/trip-notification.service.js:56]: Ensure you add a test for this line and line 63

File: src/tests/notification/socketio.test.js [hunk 32]
```diff
@@ -0,0 +1,110 @@
+/* eslint-disable no-unused-vars */
```
Comment (adesege) [inline src/tests/notification/socketio.test.js]: Remove this file from the test coverage report

File: src/middlewares/notification.middleware.js (hunk 20 above)
Comment (victor-abz) [reply src/middlewares/notification.middleware.js]: He can have them both. In DB Bernard suggested we create two columns inApp and email preference. The values are booleans.  When a user submit an update then I need to find the row to update(filter) then check its value. if it is not equal to value from body(map) that when I continue.

File: src/seeders/20200106072811-users-notification-mode.js (hunk 24 above)
Comment (victor-abz) [reply src/seeders/20200106072811-users-notification-mode.js]: Yes because each user we create in seeds needs to have default preferences set. otherwise, it would throw an error when trying to send notifications

File: src/public/script.js [hunk 33]
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-console */
```
Comment (adesege) [inline src/public/script.js]: Do not disable any eslint rule

File: src/public/script.js [hunk 34]
```diff
@@ -0,0 +1,52 @@
...
+  messageElement.innerText = message;
+  messageContainer.append(messageElement);
+}
+const token = prompt('Provide Authorization Token');
+
+socket.on('connect', () => {
+  socket
+    .emit('authenticate', { token }) // send the jwt
+    .on('authenticated', () => {
+      // console.log(value);
```
Comment (adesege) [inline src/public/script.js]: Remove commented code

File: src/routes/accommodation.route.js [hunk 35]
```diff
@@ -1,11 +0,0 @@
-import express from 'express';
```
Comment (adesege) [inline src/routes/accommodation.route.js]: Why is this file deleted?

File: src/seeders/20200106071439-demo-users.js [hunk 36]
```diff
@@ -0,0 +1,45 @@
+import faker from 'faker';
+import BcryptService from '../services/bcrypt.service';
+
+const passwordHash = BcryptService.hashPassword('Pass1234');
+
+const demoUsers = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert('Users', [{
+    id: 31,
```
Comment (adesege) [inline src/seeders/20200106071439-demo-users.js]: The id should be auto-generated.

File: src/seeders/20200106072811-users-notification-mode.js (hunk 24 above)
Comment (adesege) [reply src/seeders/20200106072811-users-notification-mode.js]: I don't see the need to generate seeders for preferences and users tables. You can manually generate these data locally. There's no point having this in staging or production environment

File: src/public/script.js (hunk 33 above)
Comment (victor-abz) [reply src/public/script.js]: This is the file for the front-end notification test that we said we will be deleting. Because some values are gotten in the connected Html, then it is hard for me to resolve the eslint errors

File: src/seeders/20200106071439-demo-users.js (hunk 36 above)
Comment (victor-abz) [reply src/seeders/20200106071439-demo-users.js]: Because I need to use the same Id in the preference seeders, 
Especially for line 5, 13, 17 If I remove the eslint rule and then try to assign variables, then the HTML won't work.

File: src/helpers/transformErrorHandler.js [hunk 37]
```diff
@@ -0,0 +1,16 @@
+import ResponseService from '../services/response.service';
+
+const transformErrorHandler = (schema, body, res, next) => {
+  const { error } = schema.validate(body);
+
+  if (error) {
+    // const errors = [];
```
Comment (adesege) [inline src/helpers/transformErrorHandler.js]: No commented code

Comment (adesege) [reply src/helpers/transformErrorHandler.js]: Ensure you check through the codebase and remove any commented code

File: src/middlewares/auth.middleware.js [hunk 38]
```diff
@@ -11,6 +11,14 @@ const authMiddleware = {
     }
     next();
   },
+  checkIfUserHaveManager: async (req, res, next) => {
+    const { lineManagerId } = await UserService.findUserByProperty({ email: req.userData.email });
+    if (lineManagerId === null || lineManagerId === undefined) {
+      ResponseService.setError(401, 'You don\'t have manager assigned yet. Please contact Admin');
```
Comment (adesege) [inline src/middlewares/auth.middleware.js]: The status code here shouldn't be 401.

File: src/migrations/20191218075931-create-users.js [hunk 39]
```diff
@@ -64,9 +64,6 @@ export function up(queryInterface, Sequelize) {
       type: Sequelize.BOOLEAN,
       defaultValue: false
     },
-    lineManagerId: {
```
Comment (adesege) [inline src/migrations/20191218075931-create-users.js]: Why is this column removed from migration?

File: src/public/script.js [hunk 40]
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable no-alert */
```
Comment (adesege) [inline src/public/script.js]: Really not happy I have to give this feedback the 3rd time. Do not disable any eslint rule.

File: src/seeders/20200106071439-demo-users.js (hunk 36 above)
Comment (adesege) [reply src/seeders/20200106071439-demo-users.js]: Then you should look for a way to dynamically map each created user to their preference.

File: src/services/preference.service.js [hunk 41]
```diff
@@ -0,0 +1,44 @@
...
+    return Preferences.create(preference);
+  }
+
+  /**
+   * @static
+   * @param {*} property
+   * @memberof PreferenceService
+   * @return {notification} preferences
+   */
+  static findPreferences(property) {
```
Comment (adesege) [inline src/services/preference.service.js]: There is a mismatch here. The function name is in the plural form but it's only going to return one preference for a particular user.

File: src/public/script.js (hunk 33 above)
Comment (adesege) [reply src/public/script.js]: Prefix each variable with `window.` because all declared variables on the browser are scoped to the window object. Moreover, that has nothing to do with the eslint rule you disabled. You shouldn't have console.log in your code by the way.

File: src/migrations/20191218075931-create-users.js (hunk 39 above)
Comment (victor-abz) [reply src/migrations/20191218075931-create-users.js]: There were duplicates from develop. both in migration and model

File: src/public/script.js (hunk 33 above)
Comment (victor-abz) [reply src/public/script.js]: Console.log was removed. but even if I set every variable. `window.` the error with eslint still come

**Review by adesege — APPROVED** [review]
